ehrql/query_language.py

```diff
@@ -343,6 +355,17 @@ def parse_date_if_str(value):
         return value
 
 
+def harmonise_datelike_series(series):
```

Comment:
I think the trick here is use the `_cast` method because that will give us support for casting more generally (e.g. using int literals with floats). Something like:
```py
def cast_all_arguments(args):
    series_args = [arg for arg in args if isinstance(arg, BaseSeries)]
    if series_args:
        # Choose the first series arbitrarily, the type checker will enforce that they
        # are all the same type in any case
        cast = series_args[0]._cast
        return [cast(arg) for arg in args]
    else:
        # This would only be the case if all the arguments were literals, which would be
        # an unusual and pointless bit of ehrQL, but we should handle it without error
        return args
```

Reply:
Ah yes I had started down that path but got too focused on dates, good catch!